packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isFalse, isUndefined, isNull } from '@lwc/shared';
+import { VM, scheduleRehydration } from './vm';
+import { markComponentAsDirty, ComponentConstructor } from './component';
+import { Template } from './template';
+import { StylesheetFactory } from './stylesheet';
+
+const swappedTemplateMap = new WeakMap<Template, Template>();
+const swappedComponentMap = new WeakMap<ComponentConstructor, ComponentConstructor>();
+const swappedStyleMap = new WeakMap<StylesheetFactory, StylesheetFactory>();
+
+const activeTemplates = new WeakMap<Template, Set<VM>>();
+const activeComponents = new WeakMap<ComponentConstructor, Set<VM>>();
+const activeStyles = new WeakMap<StylesheetFactory, Set<VM>>();
+
+function rehydrateHotTemplate(tpl: Template): boolean {
+    const list = activeTemplates.get(tpl);
+    list?.forEach((vm) => {
+        if (isFalse(vm.isDirty)) {
+            // forcing the vm to rehydrate in the next tick
+            markComponentAsDirty(vm);
+            scheduleRehydration(vm);
+        }
+    });
+    return true;
+}
+
+function rehydrateHotStyle(style: StylesheetFactory): boolean {
+    const list = activeStyles.get(style);
+    list?.forEach((vm) => {
+        // hacky way to force the styles to get recomputed
+        // by replacing the value of old template, which is used
+        // during the rendering process. If the template returned
+        // by render() is different from the previous stored template
+        // the styles will be reset, along with the content of the
+        // shadow, this way we can guarantee that the styles will be
+        // recalculated, and applied.
+        vm.cmpTemplate = () => [];
```

Comment:
We are leaking internal details about how the engine decides to re-render or not a component here (eg. resetting the `cmpTemplate`). What do you think about exposing a method out of the `VM.ts` like `forceHydrate` to scope this?

Reply:
I'm fine with that too... will do